File: health_insurance_cross_sell_prediction.jl
### load libraries ###
using CSV, DataFrames, Dates, CategoricalArrays
df = DataFrames
ca = CategoricalArrays
using ZipFile
using Distributions, StatsBase, Statistics
sb = StatsBase
st = Statistics
using Colors, ColorBrewer, Plots, StatsPlots, PlotThemes
plt = Plots
using MLBase, MLJScientificTypes, MLJ, MLJTuning
mlj = MLJ
@load DecisionTreeClassifier pkg = "DecisionTree"
@load RandomForestClassifier pkg = "DecisionTree"
@load LogisticClassifier pkg = "MLJLinearModels"
@load XGBoostClassifier pkg = "XGBoost"
using ShapML
shap = ShapML
using Base.Threads


### set variables ###
data_location = "..."


### set functions ###
## train/test split
function sample_split(dt::DataFrames.DataFrame, percent::Float64=0.7)
    n_train = sb.sample(collect(1:1:df.nrow(dt)), Int64(floor(df.nrow(dt) * percent)); replace = false)
    n_test_df = df.DataFrame(x1=collect(1:1:df.nrow(dt)))
    n_test_df[!, :x2] = ifelse.(in.(n_test_df[:x1], (n_train, )) .== true, 0, 1)
    n_test = n_test_df[n_test_df[:x2] .== 1, :x1]
    # n_test = Int64[]
    # for i in collect(1:1:df.nrow(dt))
    #     if in(i, n_train) == false
    #         push!(n_test, i)
    #     end
    # end
    return n_train, n_test
end


### import data ###
## health train
hl_train = CSV.read(string(data_location, "train.csv"), df.DataFrame; delim=",", header=1)
hl_train[!, :Region_Code] = string.(convert.(Int64, hl_train[:, :Region_Code]))
hl_train[!, :Policy_Sales_Channel] = string.(convert.(Int64, hl_train[:, :Policy_Sales_Channel]))

## health test
hl_test = CSV.read(string(data_location, "test.csv"), df.DataFrame; delim=",", header=1)
hl_test[!, :Region_Code] = string.(convert.(Int64, hl_test[:, :Region_Code]))
hl_test[!, :Policy_Sales_Channel] = string.(convert.(Int64, hl_test[:, :Policy_Sales_Channel]))

## health submission
hl_submission = CSV.read(string(data_location, "sample_submission.csv"), df.DataFrame; delim=",", header=1)


### exploratory data analysis ###
## null counts
# train
println(df.describe(hl_train)[:, [Symbol("variable"), Symbol("nmissing")]])

# test
println(df.describe(hl_test)[:, [Symbol("variable"), Symbol("nmissing")]])

# submission
println(df.describe(hl_submission)[:, [Symbol("variable"), Symbol("nmissing")]])

## distributions
x_1 = df.by(hl_train, :Region_Code, 
    count = :id => length,
    percent = :id => x -> length(x) / df.nrow(hl_train))
df.sort!(x_1, (order(:count, rev = true), :Region_Code))
println(x_1) # keep region codes 28, 8, 46, 41, 15, 30, 29, 50, and 3. remaining can be categorised as other
x_1 = nothing

x_1 = df.by(hl_train, :Policy_Sales_Channel,
    count = :id => length,
    percent = :id => x -> length(x) / df.nrow(hl_train))
df.sort!(x_1, (order(:count, rev = true), :Policy_Sales_Channel))
println(x_1) # keep policy sales channels 152, 26, 124, 160, 156, 122, 157, 154, and 151. remaining can be categorised as other


### feature engineering ###
## scientific types
schema(hl_train)

## data analysis and transformations
# gender
df.by(hl_train, :Gender,
    count = :id => x -> length(x),
    percent = :id => x -> length(x) / df.nrow(hl_train)) # 54% male
x_1 = df.by(hl_train, :Gender, percent = :id => x -> length(x) / df.nrow(hl_train))
@df x_1 plt.bar(:Gender, :percent;
    title = "gender distribution", xlabel = "gender", ylabel = "percent",
    color = "black", fill = (0, 0.8, :orange), legend = :none)
x_1 = nothing

# age
sb.summarystats(hl_train[:, :Age]) # min 20, 1st quart. 25, median 36, 3rd quart. 49, max 85, mean 39
@df hl_train plt.density(:Age;
    title = "age distribution", xlabel = "age", ylabel = "density",
    color = :black, fill = (0, 0.8, :lightgreen), legend = :none)

# driving licence
df.by(hl_train, :Driving_License,
    count = :id => x -> length(x),
    percent = :id => x -> length(x) / df.nrow(hl_train)) # ~100% licenced

# previously insured
df.by(hl_train, :Previously_Insured,
    count = :id => x -> length(x),
    percent = :id => x -> length(x) / df.nrow(hl_train)) # 46% yes

# vehicle age
df.by(hl_train, :Vehicle_Age,
    count = :id => x -> length(x),
    percent = :id => x -> length(x) / df.nrow(hl_train)) # 43% < 1 year, 53% 1-2 years

# vehicle damage
df.by(hl_train, :Vehicle_Damage,
    count = :id => x -> length(x),
    percent = :id => x -> length(x) / df.nrow(hl_train)) # 50 damaged

# region code
hl_train[!, :Region_Code_2] = ifelse.(in.(hl_train[:Region_Code], (["28","8","46","41","15","30","29","50","3"],)) .== true, hl_train[:Region_Code], "other")
df.by(hl_train, :Region_Code_2,
    count = :id => x -> length(x),
    percent = :id => x -> length(x) / df.nrow(hl_train)) # 27% 28, 38% other

# policy sales channel
hl_train[!, :Policy_Sales_Channel_2] = ifelse.(in.(hl_train[:Policy_Sales_Channel], (["152","26","124","160","156","122","157","154","151"],)) .== true, hl_train[:Policy_Sales_Channel], "other")
df.by(hl_train, :Policy_Sales_Channel_2,
    count = :id => x -> length(x),
    percent = :id => x -> length(x) / df.nrow(hl_train)) # 35% 152, 9% other

# vintage
sb.summarystats(hl_train[:, :Vintage]) # min 10, 1st quart. 82, median 154, 3rd quart. 227, max 299, mean 154
@df hl_train plt.density(:Vintage,
    title = "vintage distribution", xlabel = "vintage", ylabel = "density",
    color = "black", fill = (0, 0.8, :magenta), legend = :none) # almost uniform distribution

# annual premium
sb.summarystats(hl_train[:, :Annual_Premium]) # min 2630, 1st quart. 24405, median 31669, 3rd quart. 39400, max 540165, mean 30564
@df hl_train plt.density(:Annual_Premium,
    title = "annual premium distribution", xlabel = "annual premium", ylabel = "density",
    color = "black", fill = (0, 0.8, :lightblue), legend = :none)

# response
df.by(hl_train, :Response,
    count = :id => x -> length(x),
    percent = :id => x -> length(x) / df.nrow(hl_train)) # 12% responded

## data types
# train
coerce!(hl_train,
    :Gender => mlj.Multiclass,
    :Age => mlj.Continuous,
    :Driving_License => mlj.OrderedFactor,
    :Region_Code_2 => mlj.Multiclass,
    :Vehicle_Age => mlj.Multiclass,
    :Vehicle_Damage => mlj.Multiclass,
    :Policy_Sales_Channel_2 => mlj.Multiclass,
    :Previously_Insured => mlj.OrderedFactor,
    :Response => mlj.OrderedFactor)
ca.levels!(hl_train[:Gender], ["Male","Female"])
ca.levels!(hl_train[:Region_Code_2], ["28","8","46","41","15","30","29","50","3","other"])
ca.levels!(hl_train[:Vehicle_Age], ["< 1 Year","1-2 Year","> 2 Years"])
ca.levels!(hl_train[:Vehicle_Damage], ["Yes","No"])
ca.levels!(hl_train[:Policy_Sales_Channel_2], ["152","26","124","160","156","122","157","154","151","other"])
hl_train[:Age_Bin] = ca.cut(hl_train[:Age], [18, 25, 30, 35, 40, 45, 50, 55, 60, 65, Inf])
hl_train[:Vintage_Bin] = ca.cut(hl_train[:Vintage], [0, 50, 100, 150, 200, 250, Inf])
hl_train[:Annual_Premium_Bin] = ca.cut(hl_train[:Annual_Premium], [0, 20000, 25000, 30000, 35000, 40000, 45000, 50000, Inf])

# test
hl_test[!, :Region_Code_2] = ifelse.(in.(hl_test[:Region_Code], (["28","8","46","41","15","30","29","50","3"],)) .== true, hl_test[:Region_Code], "other")
hl_test[!, :Policy_Sales_Channel_2] = ifelse.(in.(hl_test[:Policy_Sales_Channel], (["152","26","124","160","156","122","157","154","151"],)) .== true, hl_test[:Policy_Sales_Channel], "other")
coerce!(hl_test,
    :Gender => mlj.Multiclass,
    :Age => mlj.Continuous,
    :Driving_License => mlj.OrderedFactor,
    :Region_Code_2 => mlj.Multiclass,
    :Vehicle_Age => mlj.Multiclass,
    :Vehicle_Damage => mlj.Multiclass,
    :Policy_Sales_Channel_2 => mlj.Multiclass,
    :Previously_Insured => mlj.OrderedFactor,
    :Response => mlj.OrderedFactor)
ca.levels!(hl_test[:Gender], ["Male","Female"])
ca.levels!(hl_test[:Region_Code_2], ["28","8","46","41","15","30","29","50","3","other"])
ca.levels!(hl_test[:Vehicle_Age], ["< 1 Year","1-2 Year","> 2 Years"])
ca.levels!(hl_test[:Vehicle_Damage], ["Yes","No"])
ca.levels!(hl_test[:Policy_Sales_Channel_2], ["152","26","124","160","156","122","157","154","151","other"])
hl_test[:Age_Bin] = ca.cut(hl_test[:Age], [18, 25, 30, 35, 40, 45, 50, 55, 60, 65, Inf])
hl_test[:Vintage_Bin] = ca.cut(hl_test[:Vintage], [0, 50, 100, 150, 200, 250, Inf])
hl_test[:Annual_Premium_Bin] = ca.cut(hl_test[:Annual_Premium], [0, 20000, 25000, 30000, 35000, 40000, 45000, 50000, Inf])


### model development ###
## train/test split
# note we use relevant features
# core train dataset
hl_train_2 = df.select(hl_train, [:id, :Gender, :Age_Bin, :Driving_License, :Region_Code_2, :Previously_Insured, :Vehicle_Age, :Vehicle_Damage, :Annual_Premium_Bin, :Policy_Sales_Channel_2, :Vintage_Bin, :Response])

# core test dataset
hl_test_2 = df.select(hl_train, [:id, :Gender, :Age_Bin, :Driving_License, :Region_Code_2, :Previously_Insured, :Vehicle_Age, :Vehicle_Damage, :Annual_Premium_Bin, :Policy_Sales_Channel_2, :Vintage_Bin, :Response])

# split train
n_train, n_test = mlj.partition(1:df.nrow(hl_train_2), 0.8; shuffle=true)
hl_model_train = hl_train_2[n_train, :]
hl_model_test = hl_train_2[n_test, :]

## core training dataset
y, X = mlj.unpack(hl_model_train, ==(:Response), colname -> true)
df.select!(X, Not(:id))

# one hot encoded predictor variables
Xt = mlj.transform(mlj.fit!(mlj.machine(OneHotEncoder(drop_last = true), X)))

# search for models
# for i in collect(1:1:length(mlj.models(matching(Xt, y)))) println(mlj.models(matching(Xt, y))[i]) end

# check models
# mlj.models(matching(X, y))
# mlj.models(matching(Xt, y))

## glm
mlj.evaluate(LogisticClassifier(), Xt, y, resampling = Holdout(fraction_train = 0.7, shuffle=true), measures = [auc, cross_entropy])
mlj.evaluate(LogisticClassifier(), Xt, y, resampling = CV(nfolds = 5, shuffle = true), measures = [auc, cross_entropy])
hl_glm = mlj.machine(LogisticClassifier(), Xt, y)
mlj.fit!(hl_glm; force = true)
mlj.fitted_params(hl_glm)
mlj.report(hl_glm)

## decision tree
# mlj.info(DecisionTreeClassifier)
r1 = mlj.range(DecisionTreeClassifier(), :max_depth; values = [10, 20, 30])
r2 = mlj.range(DecisionTreeClassifier(), :min_samples_leaf; values = [25, 50])
r3 = mlj.range(DecisionTreeClassifier(), :min_samples_split; values = [25, 50])
r4 = mlj.range(DecisionTreeClassifier(), :min_purity_increase; values = [0.0, 0.01, 0.05])
dt_tune = TunedModel(;model = DecisionTreeClassifier(),
    tuning = Grid(),
    resampling = Holdout(fraction_train = 0.7, shuffle = true),
    measure = [auc, cross_entropy],
    repeats = 1,
    range = [r1, r2, r3, r4])
hl_dt = mlj.machine(dt_tune, Xt, y)
println(now())
mlj.fit!(hl_dt; force = true)
(r1, r2, r3, r4) = (nothing, nothing, nothing, nothing)
println(now())
mlj.save(string(data_location, "hl_dt.jlso"), hl_dt)
mlj.fitted_params(hl_dt)
mlj.report(hl_dt).best_model
mlj.report(hl_dt).best_result
mlj.report(hl_dt).best_report
mlj.report(hl_dt).history
mlj.report(hl_dt).plotting

## random RandomForestClassifier
# mlj.info(RandomForestClassifier)
r1 = mlj.range(RandomForestClassifier(), :max_depth; values = [10, 20])
r2 = mlj.range(RandomForestClassifier(), :min_samples_leaf; values = [25, 50])
r3 = mlj.range(RandomForestClassifier(), :min_samples_split; values = [25, 50])
r4 = mlj.range(RandomForestClassifier(), :min_purity_increase; values = [0.00, 0.05])
r5 = mlj.range(RandomForestClassifier(), :n_trees; values = [50, 100, 150])
rf_tune = TunedModel(; model = RandomForestClassifier(),
    tuning = Grid(),
    resampling = Holdout(fraction_train = 0.7, shuffle = true),
    measure = [auc, cross_entropy],
    repeats = 1,
    range = [r1, r2, r3, r4, r5])
hl_rf = mlj.machine(rf_tune, Xt, y)
println(now())
mlj.fit!(hl_rf; force = true)
(r1, r2, r3, r4, r5) = (nothing, nothing, nothing, nothing, nothing)
println(now())
mlj.save(string(data_location, "hl_rf.jlso"), hl_rf)
mlj.fitted_params(hl_rf)
mlj.report(hl_rf).best_model
mlj.report(hl_rf).best_result
mlj.report(hl_rf).best_report
mlj.report(hl_rf).history
mlj.report(hl_rf).plotting

File: health_insurance_cross_sell_prediction.R
### load libraries ###
library(lubridate)
library(data.table)
library(scales)
library(RColorBrewer)
library(ggplot2)
library(cowplot)
library(caret)
library(gbm)
library(pROC)
library(mmpf)
library(parallel)


### set variables ###
data_location <- "..."
r_func_location <- "..."
no_cores <- 1


### set functions ###
source(file=paste0(r_func_location, "profiler.R"))


### import data ###
## health train
hl_train <- fread(file=paste0(data_location, "train.csv"), sep=",", header=TRUE, check.names=FALSE, stringsAsFactors=FALSE)
setDT(hl_train)

## health test
hl_test <- fread(file=paste0(data_location, "test.csv"), sep=",", header=TRUE, check.names=FALSE, stringsAsFactors=FALSE)
setDT(hl_test)

## health submission sample
hl_sub <- fread(file=paste0(data_location, "sample_submission.csv"), sep=",", header=TRUE, check.names=FALSE, stringsAsFactors=FALSE)
setDT(hl_sub)


### feature engineering ###
## health train
hl_train[, Gender:=factor(x=Gender, levels=c("Male","Female"))]
hl_train[, Age_Bin:=cut(x=Age, breaks=c(18, seq(from=25, to=65, by=5), Inf), right=FALSE)]
hl_train[, Region_Code_2:=as.character(Region_Code)]
hl_train[, Region_Code_2:=ifelse(Region_Code_2 %in% c("28","8","46","41","15","30","29","50","3"), Region_Code_2, "other")]
hl_train[, Region_Code_2:=factor(x=Region_Code_2, levels=c("28","8","46","41","15","30","29","50","3","other"))]
hl_train[, Vehicle_Age:=factor(x=Vehicle_Age, levels=c("< 1 Year","1-2 Year","> 2 Years"))]
hl_train[, Vehicle_Damage:=factor(x=Vehicle_Damage, levels=c("Yes","No"))]
hl_train[, Annual_Premium_Bin:=cut(x=Annual_Premium, breaks=c(0, seq(from=20000, to=50000, by=5000), Inf), right=FALSE)]
hl_train[, Policy_Sales_Channel_2:=as.character(Policy_Sales_Channel)]
hl_train[, Policy_Sales_Channel_2:=ifelse(Policy_Sales_Channel_2 %in% c("152","26","124","160","156","122","157","154","151"), Policy_Sales_Channel_2, "other")]
hl_train[, Policy_Sales_Channel_2:=factor(x=Policy_Sales_Channel_2, levels=c("152","26","124","160","156","122","157","154","151","other"))]
hl_train[, Vintage_Bin:=cut(x=Vintage, breaks=c(seq(from=0, to=250, by=50), Inf), right=FALSE)]
hl_train[, Response:=factor(x=Response, levels=c(0,1))]

## health test
hl_test[, Gender:=factor(x=Gender, levels=c("Male","Female"))]
hl_test[, Age_Bin:=cut(x=Age, breaks=c(18, seq(from=25, to=65, by=5), Inf), right=FALSE)]
hl_test[, Region_Code_2:=as.character(Region_Code)]
hl_test[, Region_Code_2:=ifelse(Region_Code_2 %in% c("28","8","46","41","15","30","29","50","3"), Region_Code_2, "other")]
hl_test[, Region_Code_2:=factor(x=Region_Code_2, levels=c("28","8","46","41","15","30","29","50","3","other"))]
hl_test[, Vehicle_Age:=factor(x=Vehicle_Age, levels=c("< 1 Year","1-2 Year","> 2 Years"))]
hl_test[, Vehicle_Damage:=factor(x=Vehicle_Damage, levels=c("Yes","No"))]
hl_test[, Annual_Premium_Bin:=cut(x=Annual_Premium, breaks=c(0, seq(from=20000, to=50000, by=5000), Inf), right=FALSE)]
hl_test[, Policy_Sales_Channel_2:=as.character(Policy_Sales_Channel)]
hl_test[, Policy_Sales_Channel_2:=ifelse(Policy_Sales_Channel_2 %in% c("152","26","124","160","156","122","157","154","151"), Policy_Sales_Channel_2, "other")]
hl_test[, Policy_Sales_Channel_2:=factor(x=Policy_Sales_Channel_2, levels=c("152","26","124","160","156","122","157","154","151","other"))]
hl_test[, Vintage_Bin:=cut(x=Vintage, breaks=c(seq(from=0, to=250, by=50), Inf), right=FALSE)]


### model development ###
## profiler analysis
prof_var <- c("Gender","Age_Bin","Driving_License","Region_Code_2","Previously_Insured","Vehicle_Age","Vehicle_Damage","Annual_Premium_Bin","Policy_Sales_Channel_2","Vintage_Bin")
hl_prof <- profiler(dt_pop_1=hl_train[Response=="1"], dt_pop_2=hl_train[Response=="0"], variables=prof_var, no_cores=no_cores)
rm(prof_var)
hl_prof[, .(information_value=sum(information_value)), keyby=attribute][order(-information_value)]
saveRDS(object=hl_prof, file=paste0(data_location, "hl_prof.Rds"))

## train/test split on existing data
n_index <- sample(x=1:nrow(hl_train), size=floor(nrow(hl_train)*0.7), replace=FALSE)
hl_train_tmp <- hl_train[n_index]
hl_train_tmp[, c("id","Age","Region_Code","Vintage","Policy_Sales_Channel"):=NULL]
hl_train_tmp[, Response:=factor(x=ifelse(Response=="1", "yes", "no"), levels=c("no","yes"))]
hl_test_tmp <- hl_train[-n_index]
hl_test_tmp[, c("id","Age","Region_Code","Vintage","Policy_Sales_Channel"):=NULL]
hl_test_tmp[, Response:=factor(x=ifelse(Response=="1", "yes", "no"), levels=c("no","yes"))]
saveRDS(object=n_index, file=paste0(data_location, "n_index.Rds"))
saveRDS(object=hl_train_tmp, file=paste0(data_location, "hl_train_tmp.Rds"))
saveRDS(object=hl_test_tmp, file=paste0(data_location, "hl_train_tmp.Rds"))

names(getModelInfo())

## model training control
# training control
ctrl <- trainControl(method="boot",
                     number=1,
                     p=0.8,
                     summaryFunction=twoClassSummary,
                     classProbs=TRUE,
                     verboseIter=TRUE,
                     returnData=FALSE)

## glm
# parameter lookup (we all know there's no such here ;))
modelLookup(model="glm")

# model training
print(Sys.time())
start_time <- proc.time()
hl_glm <- train(x=hl_train_tmp[, !c("Response")], y=hl_train_tmp[, Response],
                method="glm",
                trControl=ctrl,
                family=binomial,
                model=FALSE,
                metric="ROC")
end_time <- proc.time()
print(paste0("seconds to run models: ", (end_time-start_time)[3]))

# initial model performance statistics
print(hl_glm)
summary(object=hl_glm)
varImp(object=hl_glm)

# remove additional parameters
rm(start_time, end_time)

saveRDS(object=hl_glm, file=paste0(data_location, "hl_glm.Rds"))

## gbm
# parameter lookup
modelLookup(model="gbm")

# hyper-parameters for tuning
xpnd_grid <- expand.grid(n.trees=c(600),
                         interaction.depth=c(4,6,8),
                         shrinkage=c(0.05,0.1,0.2),
                         n.minobsinnode=c(20,50))

# model training
print(Sys.time())
start_time <- proc.time()
hl_gbm <- train(x=hl_train_tmp[, !c("Response")], y=hl_train_tmp[, Response],
                method="gbm",
                trControl=ctrl,
                tuneGrid=xpnd_grid,
                verbose=FALSE,
                metric="ROC")
end_time <- proc.time()
print(paste0("seconds to run models: ", (end_time-start_time)[3]))

# initial model performance statistics
print(hl_gbm)
summary(object=hl_gbm, plotit=FALSE)
varImp(hl_gbm)

# remove additional parameters
rm(xpnd_grid, start_time, end_time)

saveRDS(object=hl_gbm, file=paste0(data_location, "hl_gbm.Rds"))

## xgboost
# parameter lookup
modelLookup("xgbTree")

# hyper-parameters for tuning
xpnd_grid <- expand.grid(nrounds=500,
                         max_depth=c(4,6,8),
                         eta=c(0.05,0.1,0.2),
                         gamma=c(0,1),
                         colsample_bytree=c(1),
                         min_child_weight=c(1),
                         subsample=c(1))

# model training
print(Sys.time())
start_time <- proc.time()
hl_train_tmp_2 <- as.matrix(model.matrix(~., data=hl_train_tmp[, !c("Response")])[, -1])
hl_xgb <- train(x=hl_train_tmp_2, y=hl_train_tmp[, Response],
                method="xgbTree",
                trControl=ctrl,
                tuneGrid=xpnd_grid,
                verbose=0,
                metric="ROC")
end_time <- proc.time()
print(paste0("seconds to run models: ", (end_time-start_time)[3]))

# initial model performance statistics
print(hl_xgb)
summary(object=hl_xgb)
varImp(hl_xgb)

# remove additional parameters
rm(hl_train_tmp_2, xpnd_grid, start_time, end_time)

saveRDS(object=hl_xgb, file=paste0(data_location, "hl_xgb.Rds"))

## random forest
# parameter lookup
modelLookup("rf")

# hyper-parameters for tuning
xpnd_grid <- expand.grid(mtry=c(2,3,4,5))

# model training
print(Sys.time())
start_time <- proc.time()
hl_rf <- train(x=hl_train_tmp[, !c("Response")], y=hl_train_tmp[, Response],
               method="rf",
               trControl=ctrl,
               tuneGrid=xpnd_grid,
               ntrees=200,
               importance=TRUE,
               metric="ROC")
end_time <- proc.time()
print(paste0("seconds to run models: ", (end_time-start_time)[3]))

# initial model performance statistics
print(hl_rf)
summary(object=hl_rf)
varImp(object=hl_rf)

# remove additional parameters
rm(xpnd_grid, start_time, end_time)

saveRDS(object=hl_rf, file=paste0(data_location, "hl_rf.Rds"))

## neural network
# parameter lookup
modelLookup("nnet")

# hyper-parameters for tuning
xpnd_grid <- expand.grid(size=c(3,5,8,10,15,20),
                         decay=c(0,0.5,1))

# model training
print(Sys.time())
start_time <- proc.time()
hl_nn <- train(x=hl_train_tmp[, !c("Response")], y=hl_train_tmp[, Response],
               method="nnet",
               trControl=ctrl,
               tuneGrid=xpnd_grid,
               maxit=150,
               metric="ROC")
end_time <- proc.time()
print(paste0("seconds to run models: ", (end_time-start_time)[3]))

# initial model performance statistics
print(hl_nn)
summary(object=hl_nn)
varImp(object=hl_nn)

# remove additional parameters
rm(xpnd_grid, start_time, end_time)

saveRDS(object=hl_nn, file=paste0(data_location, "hl_nn.Rds"))


### further model testing ###
## roc on second test dataset
par(pty="s")
roc(response=hl_test_tmp[, ifelse(Response=="yes", 1, 0)],
    predictor=predict(object=hl_glm, newdata=hl_test_tmp, type="prob")[, 2],
    plot=TRUE, col="green", lwd=2.5, legacy.axes=TRUE, print.auc=TRUE,
    main="ROC Analysis", xlab="FPR", ylab="TPR")
plot.roc(x=hl_test_tmp[, ifelse(Response=="yes", 1, 0)],
         predictor=predict(object=hl_gbm, newdata=hl_test_tmp, type="prob")[, 2],
         col="royalblue", lwd=2.5, add=TRUE, print.auc=TRUE, print.auc.y=0.45)
plot.roc(x=hl_test_tmp[, ifelse(Response=="yes", 1, 0)],
         predictor=predict(object=hl_xgb, newdata=as.matrix(model.matrix(~., data=hl_test_tmp[, !c("Response")])[, -1]), type="prob")[, 2],
         col="coral", lwd=2.5, add=TRUE, print.auc=TRUE, print.auc.y=0.40)
plot.roc(x=hl_test_tmp[, ifelse(Response=="yes", 1, 0)],
         predictor=predict(object=hl_rf, newdata=hl_test_tmp, type="prob")[, 2],
         col="purple", lwd=2.5, add=TRUE, print.auc=TRUE, print.auc.y=0.35)
plot.roc(x=hl_test_tmp[, ifelse(Response=="yes", 1, 0)],
         predictor=predict(object=hl_nn, newdata=hl_test_tmp, type="prob")[, 2],
         col="orange", lwd=2.5, add=TRUE, print.auc=TRUE, print.auc.y=0.30)
legend("bottomright", legend=c("logistic regression","gradient boosting","xgboost","random forest","neural network"), col=c("green","royalblue","coral","purple","orange"), cex=0.75, lwd=2.5, bty="n")
par(pty="m")

## optimal number of trees gbm
hl_gbm_tr <- lapply(1:hl_gbm$finalModel$n.trees,
                    function(x1) {
                        if (x1 ==1 || x1 %% 10 == 0) {
                            print(paste0("number of trees: ", x1))
                        }
                        
                        val <- hl_test_tmp[, ifelse(Response=="yes", 1, 0)]
                        pred <- predict(object=hl_gbm$finalModel, newdata=model.matrix(~., data=hl_test_tmp)[, -1], n.trees=x1, type="response")
                        
                        dt_1_roc <- roc(response=val, predictor=pred)
                        
                        dt_1 <- data.table(n_trees=x1,
                                           auc=auc(dt_1_roc))
                        
                        return(dt_1)
                    })
hl_gbm_tr <- rbindlist(l=hl_gbm_tr, use.names=TRUE, fill=TRUE)
